caiman/motion_correction.py

```diff
@@ -2098,6 +2243,272 @@ def tile_and_correct(img, template, strides, overlaps, max_shifts, newoverlaps=N
+
+        border_nan : bool or string, optional
+            specifies how to deal with borders. (True, False, 'copy', 'min')
+
+    Returns:
+        (new_img, total_shifts, start_step, xyz_grid)
+            new_img: ndarray, corrected image
+
+
+    """
+
+    img = img.astype(np.float64).copy()
+    template = template.astype(np.float64).copy()
+
+    if gSig_filt is not None:
+
+        img_orig = img.copy()
+        img = high_pass_filter_space(img_orig, gSig_filt)
+
+    img = img + add_to_movie
+    template = template + add_to_movie
+
+    # compute rigid shifts
+    rigid_shts, sfr_freq, diffphase = register_translation_3d(
+        img, template, upsample_factor=upsample_factor_fft, max_shifts=max_shifts)
+
+    if max_deviation_rigid == 0: # if rigid shifts only
+
+#        if shifts_opencv:
+            # NOTE: opencv does not support 3D operations - skimage is used instead
+ #       else:
+
+        if gSig_filt is not None:
+            raise Exception(
+                'The use of FFT and filtering options have not been tested. Set opencv=True')
+
+        new_img = apply_shifts_dft(
+            sfr_freq, (-rigid_shts[0], -rigid_shts[1], -rigid_shts[2]), diffphase, border_nan=border_nan)
+
+        return new_img - add_to_movie, (-rigid_shts[0], -rigid_shts[1]), None, None
+    else:
+        # extract patches
+        templates = [
+            it[-1] for it in sliding_window_3d(template, overlaps=overlaps, strides=strides)]
+        xyz_grid = [(it[0], it[1], it[2]) for it in sliding_window_3d(
+            template, overlaps=overlaps, strides=strides)]
+        num_tiles = np.prod(np.add(xyz_grid[-1], 1))
+        imgs = [it[-1]
+                for it in sliding_window_3d(img, overlaps=overlaps, strides=strides)]
+        dim_grid = tuple(np.add(xyz_grid[-1], 1))
+
+        if max_deviation_rigid is not None:
+
+            lb_shifts = np.ceil(np.subtract(
+                rigid_shts, max_deviation_rigid)).astype(int)
+            ub_shifts = np.floor(
+                np.add(rigid_shts, max_deviation_rigid)).astype(int)
+
+        else:
+
+            lb_shifts = None
+            ub_shifts = None
+
+        # extract shifts for each patch
+        shfts_et_all = [register_translation_3d(
+            a, b, c, shifts_lb=lb_shifts, shifts_ub=ub_shifts, max_shifts=max_shifts) for a, b, c in zip(
+            imgs, templates, [upsample_factor_fft] * num_tiles)]
+        shfts = [sshh[0] for sshh in shfts_et_all]
+        diffs_phase = [sshh[2] for sshh in shfts_et_all]
+        # create a vector field
+        shift_img_x = np.reshape(np.array(shfts)[:, 0], dim_grid)
+        shift_img_y = np.reshape(np.array(shfts)[:, 1], dim_grid)
+        shift_img_z = np.reshape(np.array(shfts)[:, 2], dim_grid)
+        diffs_phase_grid = np.reshape(np.array(diffs_phase), dim_grid)
+
+        #  shifts_opencv doesn't make sense here- replace with shifts_skimage
+        if shifts_opencv:
```

Comment:
I would focus only `shifts_opencv = True`, at least for the beginning. It's simpler, faster (at least in 2d) and can easily handled motion fields upsampled to the original resolution. The only downside is that linear or cubic interpolation introduces a tiny amount of spatial smoothing that FFT interpolation does not. But I don't think this will play any role in practice.